ceremonies/src/lib.rs

```diff
@@ -358,147 +375,465 @@ decl_error! {
-			for p in 1..=pcount {
-				let participant = <ParticipantRegistry<T>>::get((cid, cindex), &p);
-				if Self::participant_reputation((cid, cindex), &participant) ==
-					Reputation::UnverifiedReputable
-				{
-					reputables.push(participant);
-				} else if <encointer_communities::Module<T>>::bootstrappers(cid)
-					.contains(&participant)
-				{
-					bootstrappers.push(participant);
-				} else if <Endorsees<T>>::contains_key((cid, cindex), &participant) {
-					endorsees.push(participant)
-				} else {
-					newbies.push(participant);
-				}
+			Self::generate_meetup_assignment_params((*cid, cindex)).ok();
+		}
+		Ok(())
+	}
+
+	fn is_prime(n: u64) -> bool {
+		if n <= 3 {
+			return n > 1
+		}
+		if n % 2 == 0 || n % 3 == 0 {
+			return false
+		}
+		if n < 25 {
+			return true
+		}
+		let mut i: u64 = 5;
+		while i.pow(2) <= n {
+			if n % i == 0u64 || n % (i + 2u64) == 0u64 {
+				return false
 			}
+			i += 6u64;
+		}
+		return true
+	}
 
-			// n == amount of valid registrations that fit into the meetups
-			let mut n = bootstrappers.len() + reputables.len();
-			n += min(newbies.len(), n / 2);
-			n += endorsees.len();
+	fn find_prime_below(mut n: u64) -> u64 {
+		if n <= 2 {
+			return 2u64
+		}
+		if n % 2 == 0 {
+			n -= 1;
+		}
+		while n > 0 {
+			if Self::is_prime(n) {
+				return n
+			}
+			n -= 2;
+		}
+		2u64
+	}
 
-			if n < 3 {
-				debug!(target: LOG, "no meetups assigned for cid {:?}", cid);
-				continue
+	fn mod_inv(a: i64, module: i64) -> i64 {
+		let mut mn = (module, a);
+		let mut xy = (0, 1);
+
+		while mn.1 != 0 {
+			xy = (xy.1, xy.0 - (mn.0 / mn.1) * xy.1);
+			mn = (mn.1, mn.0 % mn.1);
+		}
+
+		while xy.0 < 0 {
+			xy.0 += module;
+		}
+		xy.0
+	}
+
+	fn validate_equal_mapping(
+		num_participants: u64,
+		assignment_params: AssignmentParams,
+		n: u64,
+	) -> bool {
+		if num_participants < 2 {
+			return true
+		}
+
+		let mut meetup_index_count: Vec<u64> = vec![0; n as usize];
+		let meetup_index_count_max = ceil_division(num_participants - assignment_params.m, n);
+		for i in assignment_params.m..num_participants {
+			let meetup_index = Self::assignment_fn(i, assignment_params, n);
+			meetup_index_count[meetup_index as usize] += 1;
+			if meetup_index_count[meetup_index as usize] > meetup_index_count_max {
+				return false
 			}
+		}
+		true
+	}
 
-			// ensure that every meetup has at least one experienced participant
-			n = min(n, (bootstrappers.len() + reputables.len()) * 12);
+	fn get_random_nonzero_group_element(m: u64) -> u64 {
+		let mut random_source =
+			RandomNumberGenerator::<T::Hashing>::new(T::RandomnessSource::random_seed().0);
 
-			// capping the amount a participants prevents assigning more meetups than there are locations.
-			if n > n_locations * 12 {
-				warn!(target: LOG, "Meetup Locations exhausted for cid: {:?}", cid);
-				n = n_locations * 12;
+		// random number in [1, m-1]
+		(random_source.pick_usize((m - 2) as usize) + 1) as u64
```

Comment:
I think, we can just implement a `pick_u64` on the random number generator to make this more beautiful.